sdks/python/apache_beam/options/pipeline_options.py

```diff
@@ -348,12 +348,15 @@ def get_all_options(
         del result[k]
 
     if overrides:
-      _LOGGER.warning("Discarding invalid overrides: %s", overrides)
+      if retain_unknown_options:
+        result.update(overrides)
```

Comment:
Above, we check:

```
      if (drop_default and parser.get_default(k) == result[k] and
          not isinstance(parser.get_default(k), ValueProvider)):
        del result[k]
```

Do we need to do something similar here?

If so (or maybe regardless), it might be cleaner to conditionally do something like `for k in overrides` instead of `for k in list(result)` above when retain_unkown_options is set. That also keeps any future updates in one place

Reply:
The difficulty is that the set of pipeline options is actually the union of what's in result and what's in overrides. Also, we don't necessarily know what the defaults are for values that don't come from flags.

Reply:
> Also, we don't necessarily know what the defaults are for values that don't come from flags.

I guess my point is, will we ever know what the defaults are? If so, its probably worth keeping?

I still think we could consolidate - even though the set of pipeline options is the union of result and overrides, we're only adding the overrides in the previous loop so I think `for k in overrides` would be valid in this case.

Regardless, its at most a small code cleanliness win (if that), so I'm good leaving it as is.

Reply:
Correct, we need to keep things if we don't know (for sure) what the default is. 

We can't easily consolidate because as well as things in `overrides` that don't work for `parser.get_default` (or at least are not safe to use the guessed default) there may be things already in `result` that are not in `overrides` that we want to remove because they have default values.

I agree this code is all very messy. Were I to do it again I would probably try better to detangle the argument parsing code from PipelineOptions.